FM-68: Add support for create OpenMRS Condition through FHIR

corneliouzbett opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'SYNCT-250 Add Cohort synchronization - REST' -->
<!--- 'SYNCT-JiraIssueNumber JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->
Add support for create OpenMRS Condition through FHIR

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/FM-68

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

dkayiwa commented:
Can you put some text in the ticket description?

dkayiwa commented:
It would also be great to add some unit tests

corneliouzbett commented:
Will add unit tests

joeldenning commented on omod/src/test/java/org/openmrs/module/fhir/resources/FHIRConditionResourceTest.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Test
+    public void getByUniqueId() {
```
This test doesn't seem to do anything. Are the unit tests here To-Do / in progress?

corneliouzbett replied:
They are Todos

joeldenning replied:
Ah okay so we should hold off on reviewing/merging this PR until you finish the To-Do's?

jonathandick replied:
There is a lot of code being pushed of which I think some is ready for review. @corneliouzbett, please let us know if there's any code that you would like reviewed at this time.

jonathandick commented on api/src/main/java/org/openmrs/module/fhir/api/util/FHIRConditionUtil.java:
```diff
@@ -69,7 +74,53 @@
+			List<Coding> codings = conceptDt.getCoding();
+
+			//Set concept
+			Concept openMrsConcept = new Concept();
```
As discussed on slack, the request to create a new condition should throw an exception if the concept does not already exist in the openmrs concept dictionary. We should not be creating new concepts here.

corneliouzbett replied:
Noted

joeldenning commented:
Is this now mergeable / ready for review?

joeldenning commented on api-1.9/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl1_9.java:
```diff
@@ -0,0 +1,73 @@
+
+@Component(value = "fhir.ConditionHelper")
+@OpenmrsProfile(openmrsPlatformVersion = "1.9.*")
+public class ConditionHelperImpl1_9 implements ConditionHelper {
```
Why do we need a 1.9.x impl as well as a 1.11.x impl? I get why 1.x versus 2.x would be a big deal, but not why 1.9 versus 1.11 would be a big deal.

Asking out of curiosity

dkayiwa replied:
Very good point @joeldenning 👍

corneliouzbett replied:
@joeldenning @dkayiwa 
You are right. there's no difference between 1.11.x and 1.9.x but there's a huge difference in version 2.0.x and 2.2.x. 
That's why you find a class implementation for 1.11, 1.9 and 2.0 are just skeleton but the actual code is in the FHIRConditionUtil class.
This is how Allergy and Drug were implemented

dkayiwa replied:
If there is no difference, why not just combine them into one class, with an openmrsPlatformVersion range, instead of duplicating the same exact code in two classes?

corneliouzbett replied:
<img width="212" alt="Screenshot 2019-10-31 at 08 37 39" src="https://user-images.githubusercontent.com/19473115/67921777-e3bf5700-fbb9-11e9-8cb1-dbd7009328a7.png">
 This will be great.
@dkayiwa  where do you suggest to put the new class (which combines 1.11, 1.9, 2.0), since the previous implementation has different API modules for different versions as you can see from the above image.
Will it be confusing to have a one class with openmrsPlatformVersion range `@OpenmrsProfile(openmrsPlatformVersion = "1.9.* - 2.1.*")`  at api-2.0? It will be inconsistent.

What you mean is the previous implementation wasn't done in the right way. 
compare this two implementation [FHIRAllergyIntoleranceUtil1_9.java](https://github.com/openmrs/openmrs-module-fhir/blob/master/api-1.9/src/main/java/org/openmrs/module/fhir/util/FHIRAllergyIntoleranceUtil1_9.java) and [FHIRAllergyIntoleranceUtil1_11.java](https://github.com/openmrs/openmrs-module-fhir/blob/master/api-1.11/src/main/java/org/openmrs/module/fhir/util/FHIRAllergyIntoleranceUtil1_11.java) . To me they are duplicate. Maybe there was a reason for this or none. I don't know

@dkayiwa need your advice on this

dkayiwa commented on api-1.11/src/main/java/org/openmrs/module/fhir/util/FHIRConditionUtil1_11.java:
```diff
@@ -0,0 +1,20 @@
+ */
+package org.openmrs.module.fhir.util;
+
+public class FHIRConditionUtil1_11 {
```
What is the use of this empty class?

dkayiwa commented on api-1.9/src/main/java/org/openmrs/module/fhir/util/FHIRConditionUtil1_9.java:
```diff
@@ -0,0 +1,20 @@
+
+public class FHIRConditionUtil1_9 {
+	/**
+	 *  Implement version 1.9 FHIRCondition Util methods
```
What is the use of this empty class?

dkayiwa commented on api-2.0/src/main/java/org/openmrs/module/fhir/util/FHIRConditionUtil2_0.java:
```diff
@@ -0,0 +1,20 @@
+
+public class FHIRConditionUtil2_0 {
+	/**
+	 *  Implement version 1.11 FHIRCondition Util methods
```
What is the use of this empty class?

dkayiwa commented on api-2.0/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java:
```diff
@@ -64,4 +74,52 @@ public void getConditionByObsId_shouldReturnMatchingFHIRCondition() {
 	}
+
+	@Test
+	public void createFHIRCondition() {
```
Can you use the same naming convention as for the second test?

dkayiwa commented on api-2.0/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java:
```diff
@@ -64,4 +74,52 @@ public void getConditionByObsId_shouldReturnMatchingFHIRCondition() {
+		Context.getAdministrationService().saveGlobalProperty(new GlobalProperty(FHIRConstants
+				.CONCEPTS_CONVERTABLE_TO_CONDITIONS_STORED_AS_OBS, GLOBAL_PROPS_CONDITION_MAPPING_CONCEPT_ID));
+		Condition condition = new Condition();
+		String uuid = "75f5b373-5065-11de-80cb-001e378eb67f";
```
Did you take a look at our naming conventions for constants? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-NamingConventions

dkayiwa replied:
Did you see the above?

dkayiwa replied:
@corneliouzbett did you see the above question?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/ConditionService.java:
```diff
@@ -16,9 +16,11 @@
 
 import java.util.List;
 
+@Transactional
```
Our convention is to put the transactional annotation on the implementation instead of interface. https://tickets.openmrs.org/browse/TRUNK-208

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/helper/ConditionHelper.java:
```diff
@@ -0,0 +1,15 @@
+package org.openmrs.module.fhir.api.helper;
```
Can you add the license header?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/ConditionServiceImpl.java:
```diff
@@ -18,16 +18,15 @@
 import java.util.List;
 
-public class ConditionServiceImpl implements ConditionService {
+public class ConditionServiceImpl extends BaseOpenmrsService implements ConditionService {
```
What was the reason for extending BaseOpenmrsService?

corneliouzbett replied:
Just confirmed I don't need at all

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/ConditionServiceImpl.java:
```diff
@@ -49,27 +48,17 @@ public void setDao(FHIRDAO dao) {
 	@Override
 	public List<Condition> searchConditionById(String id) {
-		org.openmrs.Condition condition = Context.getService(org.openmrs.api.ConditionService.class)
-				.getConditionByUuid(id);
```
Why did you replace this with null?

corneliouzbett replied:
It was the previous implementation, not working as expected.
Will redo in https://issues.openmrs.org/browse/FM-65

dkayiwa replied:
Shouldn't we throw NotImplementedException instead of returning null?

dkayiwa replied:
Did you see the above?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/ConditionStrategy.java:
```diff
@@ -0,0 +1,62 @@
+
+	@Override
+	public List<Condition> searchConditionById(String uuid) {
+		return null;
```
Are you intentionally returning null in these two methods?

corneliouzbett replied:
yeah, I will work on them in the next ticket. I am only handling create method and fixing getbyId
https://issues.openmrs.org/browse/FM-65

dkayiwa replied:
Shouldn't we throw NotImplementedException instead of returning null?

dkayiwa replied:
I mean throwing NotImplementedException instead of returning null

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/GenericConditionStrategy.java:
```diff
@@ -27,7 +27,16 @@
+
+	/**
+	 * Create FHIRCondition
+	 * @param condition
```
Can you document the above param?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/ObsConditionStrategy.java:
```diff
@@ -32,6 +32,7 @@
 import java.util.Collection;
 import java.util.List;
 
+@Deprecated
```
Are you deprecating a class and at the same time adding new methods to it?

dkayiwa replied:
Did you see the above question?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/util/FHIRConditionUtil.java:
```diff
@@ -65,11 +81,243 @@
+	 *  Joins string in a list
+	 * @param separator
+	 * @param input
+	 * @return string of list items separated by separator specified in the param (commas)
```
Doesn't Java already have a standard way of doing this without us having to create and maintain a custom method?

dkayiwa commented on omod/src/main/java/org/openmrs/module/fhir/resources/FHIRConditionResource.java:
```diff
@@ -40,6 +40,7 @@ public Condition getByUniqueId(IdType id) {
 	 * @param id id of the OpenMrs Obs
 	 * @return
 	 */
+	@Deprecated
```
Deprecating a class and at the same time adding a new method to it?

corneliouzbett replied:
The same class I am deprecating extends the GenericConditionStrategy interface which I am using in condition strategy. Maybe I should delete the class since it wasn't working and It's not useful at all. I have just made it abstract.

dkayiwa replied:
If we do not need it, then get rid of it.

dkayiwa replied:
Did you see the above?

dkayiwa commented on omod/src/main/resources/config.xml:
```diff
@@ -26,14 +26,18 @@
 	</description>
 	<updateURL>https://modules.openmrs.org/modules/download/${project.parent.artifactId}/update.rdf</updateURL>
-	<require_version>2.0.5 - 2.0.*, 1.11.3, 1.9.10 - 1.9.*</require_version>
+	<require_version>2.2.*, 2.0.5 - 2.0.*, 1.11.3, 1.9.10 - 1.9.*</require_version>
```
Since platform 2.3 is getting released, can we add support for it too? And platform 2.4 which is the master branch. This way people can tests the module on master.

corneliouzbett commented:
@dkayiwa I have updated the PR, kindly review

dkayiwa commented on api-1.11/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl1_11.java:
```diff
@@ -0,0 +1,87 @@
+@OpenmrsProfile(openmrsVersion = "1.11.*")
+public class ConditionHelperImpl1_11 implements ConditionHelper {
+
+	/**
```
Instead of repeating these methods so many times with the exact same implementation, can't we put this in some sort of base class?

corneliouzbett commented:
@dkayiwa I have updated the PR.
I moved all the logic to the base class `BaseConditionImpl` then implementation class extends it.
This was great.
Should we create a ticket for the other classes with duplicate code to use base Class :(

dkayiwa commented:
@corneliouzbett can you put some text in the ticket description?

corneliouzbett commented:
@dkayiwa it's done

dkayiwa commented:
Can you throw NotImplementedException for those places where you are returning nulls simply because you have not yet done the needful?

corneliouzbett commented:
@dkayiwa done

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,182 @@
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.openmrs.module</groupId>
```
Do we need all these dependencies?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionImpl.java:
```diff
@@ -0,0 +1,61 @@
+package org.openmrs.module.fhir.api.impl.base;
```
Can we out a license header?

corneliouzbett commented:
@dkayiwa done
Remove unused dependencies
added license header for BaseConditionImpl

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,128 @@
+        </dependency>
+
+        <dependency>
+            <groupId>ca.uhn.hapi.fhir</groupId>
```
Do we need this dependency?

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,128 @@
+            <artifactId>hapi-fhir-base</artifactId>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.slf4j</groupId>
```
Do we need this dependency?

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,128 @@
+            <groupId>ca.uhn.hapi.fhir</groupId>
+            <artifactId>hapi-fhir-structures-dstu3</artifactId>
+            <exclusions>
+                <exclusion>
```
Do we need this dependency?

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,128 @@
+        <dependency>
+            <groupId>ca.uhn.hapi.fhir</groupId>
+            <artifactId>hapi-fhir-validation-resources-dstu3</artifactId>
+            <exclusions>
```
Do we need this dependency?

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,128 @@
+            <scope>test</scope>
+        </dependency>
+        
+        <dependency>
```
Do we need this dependency?

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,128 @@
+            <artifactId>phloc-schematron</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.phloc</groupId>
```
Do we need this dependency?

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,128 @@
+        </dependency>
+
+        <dependency>
+            <groupId>org.openmrs.test</groupId>
```
Do we need this dependency?

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,128 @@
+            <artifactId>openmrs-web</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.openmrs.api</groupId>
```
Do we need this dependency?

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,128 @@
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.openmrs.web</groupId>
```
A web dependency in the api layer?

dkayiwa commented on omod/src/main/resources/config.xml:
```diff
@@ -26,14 +26,18 @@
 
 	<conditionalResources>
+		<conditionalResource>
+			<path>/lib/fhir-api-2.2-${project.parent.version}.jar</path>
```
Can we make this stretch all the way to 2.4?

dkayiwa commented on omod/src/main/resources/config.xml:
```diff
@@ -228,6 +232,12 @@
 
+	<globalProperty>
+		<property>${project.parent.artifactId}.condition.strategy</property>
+		<defaultValue>DefaultConditionStrategy</defaultValue>
```
Is this just a key or a call with full package name?

corneliouzbett replied:
It's a call with full package name

dkayiwa replied:
Did you intentionally leave out the packages from the default value?

corneliouzbett replied:
I Don't understand what you mean by that.

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/helper/ConditionHelper.java:
```diff
@@ -0,0 +1,50 @@
+package org.openmrs.module.fhir.api.helper;
+
+import org.hl7.fhir.dstu3.model.Condition;
+
```
Can we document what this interface is used for?
Also add a since version annotation.

corneliouzbett replied:
done

corneliouzbett commented:
@dkayiwa Updated PR
Just realized I only needed one Dependency (Openmrs API)

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/ConditionServiceImpl.java:
```diff
@@ -47,48 +48,52 @@ public void setDao(FHIRDAO dao) {
 	@Override
-	public Condition updateFHIRCondition(Condition condition, String theId) {
+	public Condition updateFHIRCondition(Condition condition, String theId)throws NotImplementedException {
 		return null;
```
I mean throwing NotImplementedException instead of returning null

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/ConditionServiceImpl.java:
```diff
@@ -47,48 +48,52 @@ public void setDao(FHIRDAO dao) {
 	@Override
-	public List<Condition> searchConditionsByPatient(String patientUuid) {
+	public List<Condition> searchConditionsByPatient(String patientUuid) throws NotImplementedException {
 		return null;
```
I mean throwing NotImplementedException instead of returning null

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/ConditionServiceImpl.java:
```diff
@@ -47,48 +48,52 @@ public void setDao(FHIRDAO dao) {
-		}
-		return conditionList;
+	public List<Condition> searchConditionById(String id) throws NotImplementedException {
+		return null;
```
I mean throwing NotImplementedException instead of returning null

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/ConditionStrategy.java:
```diff
@@ -0,0 +1,58 @@
+	 */
+	@Override
+	public List<Condition> searchConditionByName(String name) throws NotImplementedException {
+		return null;
```
I mean throwing NotImplementedException instead of returning null

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionImpl.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class BaseConditionImpl {
```
Shouldn't this be BaseConditionHelperImpl?

dkayiwa commented on api-2.0/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java:
```diff
@@ -64,4 +89,50 @@ public void getConditionByObsId_shouldReturnMatchingFHIRCondition() {
+		Context.getAdministrationService().saveGlobalProperty(new GlobalProperty(FHIRConstants
+				.CONCEPTS_CONVERTABLE_TO_CONDITIONS_STORED_AS_OBS, GLOBAL_PROPS_CONDITION_MAPPING_CONCEPT_ID));
+		String patientUuid = "61b38324-e2fd-4feb-95b7-9e9a2a4400df";
+		String conceptUuid = "4a5048b1-cf85-4c64-9339-7cab41e5e364";
```
Did you take a look at our naming conventions for constants? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-NamingConventions

dkayiwa replied:
Do you even need these variables?

corneliouzbett replied:
Those constants were already there. Just called/refer to my code.
https://github.com/openmrs/openmrs-module-fhir/blob/master/api/src/main/java/org/openmrs/module/fhir/api/util/FHIRConstants.java
Perhaps I should rename

dkayiwa replied:
Did you look at our naming conventions for constants?

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl2_2.java:
```diff
@@ -0,0 +1,148 @@
+import static org.openmrs.module.fhir.utils.FHIRConditionUtil2_2.generateConditionClinicalStatusFromFHIRCondition;
+
+@Component(value = "fhir.ConditionHelper")
+@OpenmrsProfile(openmrsPlatformVersion = "2.2.* ")
```
Can we make this work all the way to 2.4?

dkayiwa commented on api-2.0/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java:
```diff
@@ -22,15 +41,15 @@
  */
 public class ConditionServiceTest extends BaseModuleContextSensitiveTest {
```
Do we have an equivalent of these tests for platform 2.2?

dkayiwa commented on api-2.2/pom.xml:
```diff
@@ -0,0 +1,55 @@
+    <artifactId>fhir-api-2.2</artifactId>
+    <packaging>jar</packaging>
+    <name>FHIR Module API 2.2</name>
+    <description>API project for FHIR</description>
```
Can you make this description different from the one for fhir-api?

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl2_2.java:
```diff
@@ -0,0 +1,148 @@
+	public Condition createCondition(Condition condition) {
+		List<String> errors = new ArrayList<>();
+		if (condition.isEmpty()) {
+			errors.add("Condition resource is null");
```
Shouldn't you throw an exception instead of just returning null?

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl2_2.java:
```diff
@@ -0,0 +1,148 @@
+			return null;
+		}
+		if (condition.getSubject().isEmpty()) {
+			errors.add("No subject");
```
Shouldn't you throw an exception instead of just returning null?

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/utils/FHIRConditionUtil2_2.java:
```diff
@@ -0,0 +1,97 @@
+	 * @param status condition clinical status
+	 * @return ConditionClinicalStatus
+	 */
+	public static Condition.ConditionClinicalStatus MapOpenmrsStatusToFHIRClinicalStatus(ConditionClinicalStatus status) {
```
Did you look at our convention for method names? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-NamingConventions

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl2_2.java:
```diff
@@ -0,0 +1,148 @@
+import java.util.Date;
+import java.util.List;
+
+import static org.openmrs.module.fhir.utils.FHIRConditionUtil2_2.generateConceptFromFHIRCondition;
```
Can we remove these static imports and prefix with FHIRConditionUtil2_2. ?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionImpl.java:
```diff
@@ -0,0 +1,74 @@
+	public Condition createCondition(Condition condition) {
+		List<String> errors = new ArrayList<>();
+		if (condition.isEmpty()) {
+			errors.add("Condition resource is null");
```
Can we throw an exception instead of simply returning null?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionImpl.java:
```diff
@@ -0,0 +1,74 @@
+			return null;
+		}
+		if (condition.getSubject().isEmpty()) {
+			errors.add("No subject");
```
Can we throw an exception instead of simply returning null?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionImpl.java:
```diff
@@ -0,0 +1,74 @@
+
+
+	public org.openmrs.Condition generateOpenMrsCondition(Condition condition) throws NotImplementedException {
+		return null;
```
Can we throw NotImplementedException instead of returning null?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionImpl.java:
```diff
@@ -0,0 +1,74 @@
+
+
+	public Condition generateFHIRCondition(org.openmrs.Condition condition) throws NotImplementedException {
+		return null;
```
Can we throw NotImplementedException instead of returning null?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/util/FHIRConditionUtil.java:
```diff
@@ -65,11 +81,236 @@
+	 */
+	public static org.openmrs.Condition generateOpenMRSCondition(org.hl7.fhir.dstu3.model.Condition condition) {
+		org.openmrs.Condition openMrsCondition;
+		List<String> errors = new ArrayList<>();
```
What are you eventually using the errors list for?

corneliouzbett replied:
`openMrsCondition.setConcept(FHIRUtils.getConceptFromCode(condition.getCode(), errors));`

`FHIRUtils.checkGeneratorErrorList(errors);`

dkayiwa replied:
I do not seem to see `FHIRUtils.checkGeneratorErrorList(errors)` in this method. On which line number is it?

dkayiwa replied:
@corneliouzbett did you see the above?

corneliouzbett replied:
Check Line 129

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/ConditionService.java:
```diff
@@ -27,15 +27,7 @@
-	 * @return fhir condition resource and will return null if condition is not found for the given id
-	 */
-	Condition getConditionByObsId(String id);
+	Condition getConditionById(String id);
```
If this is a uuid, can we rename the parameter accordingly?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionHelperImpl.java:
```diff
@@ -0,0 +1,82 @@
+		List<String> errors = new ArrayList<>();
+		if (condition.isEmpty()) {
+			errors.add("Condition resource is null");
+			throw new NullPointerException("Condition resource cannot be null");
```
Shouldn't this be? UnprocessableEntityException

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionHelperImpl.java:
```diff
@@ -0,0 +1,82 @@
+		}
+		if (condition.getSubject().isEmpty()) {
+			errors.add("No subject");
+			throw new NullPointerException("condition Subject cannot be null");
```
Shouldn't this be? UnprocessableEntityException

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionHelperImpl.java:
```diff
@@ -0,0 +1,82 @@
+		Obs obs = Context.getObsService()
+				.saveObs(FHIRConditionUtil.generateOpenMrsObsFromFHIRCondition(condition), "");
+
+		FHIRUtils.checkGeneratorErrorList(errors);
```
Do we still need the above after throwing exceptions on errors?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/helper/ConditionHelper.java:
```diff
@@ -0,0 +1,59 @@
+	org.openmrs.Condition generateOpenMrsCondition(Condition condition);
+
+	/**
+	 * generate FHIR condition from openmrs condition
```
Can we start the sentence with a capital letter?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/util/FHIRConditionUtil.java:
```diff
@@ -13,18 +13,35 @@
+	/**
+	 * generates FHIR Condition from openmrs condition
+	 *
+	 * @param condition fhir comdition
```
Is this param really an fhir or openmrs condition?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/util/FHIRConditionUtil.java:
```diff
@@ -65,11 +81,237 @@
-	public static org.openmrs.Condition generateOpenMRSCondition() {
-		return null;
+	/**
+	 * generates OpenmrsCondition from FHIR Conditions
```
Can we start the sentence with a capital letter?

dkayiwa replied:
Does it generate it from one condition or a list of conditions?

corneliouzbett replied:
one condition

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/util/FHIRConditionUtil.java:
```diff
@@ -65,11 +81,237 @@
+	}
+
+	/**
+	 * generates condition clinical status from FHIR condition
```
Can we start the sentence with a capital letter?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/util/FHIRConditionUtil.java:
```diff
@@ -65,11 +81,237 @@
+	}
+
+	/**
+	 * generate FHIR condition from openmrs observation
```
Can we start the sentence with a capital letter?

dkayiwa commented on omod/src/main/java/org/openmrs/module/fhir/resources/FHIRConditionResource.java:
```diff
@@ -27,34 +27,29 @@
+	/**
+	 * Create Condition
+	 *
+	 * @param condition fhir Location object
```
Location object?

dkayiwa commented on omod/src/main/java/org/openmrs/module/fhir/resources/FHIRConditionResource.java:
```diff
@@ -27,34 +27,29 @@
+	 * Create Condition
+	 *
+	 * @param condition fhir Location object
+	 * @return Method outcome object which contains the identity of the created resource.
```
What is the return value?

corneliouzbett replied:
Method outcome object which contains the identity of the created resource
`"display": "Condition was successfully created with id 7c954b1b-1d82-4cab-aa00-f14de88ecade"`

corneliouzbett replied:
It's different
`MethodOutcomeBuilder.buildCreate()` return Method outcome object unlike the other methods

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl2_2.java:
```diff
@@ -0,0 +1,146 @@
+			errors.add("Condition with given uuid doesn't exits");
+			return null;
+		}
+		FHIRUtils.checkGeneratorErrorList(errors);
```
Of what value is it when we immediately return?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/GenericConditionStrategy.java:
```diff
@@ -20,14 +20,34 @@
+	/**
+	 * Search condition by uuid
+	 *
+	 * @param uuid the uuid of the condition that need to retrieve
```
Can you correct the above grammar?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/GenericConditionStrategy.java:
```diff
@@ -20,14 +20,34 @@
+	 * Search condition by uuid
+	 *
+	 * @param uuid the uuid of the condition that need to retrieve
+	 * @return Condition Return a List of FHIR condition resource and will return Empty list if condition is not found for the given uuid
```
Did you confirm that it actually returns an empty list instead of null?

corneliouzbett replied:
It's not yet implemented.
That's search method was editing what was there.
https://issues.openmrs.org/projects/FM/issues/FM-65 this will take care of that

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/GenericConditionStrategy.java:
```diff
@@ -20,14 +20,34 @@
+	 * Search condition by name
+	 *
+	 * @param name condition name
+	 * @return Condition Return a List of FHIR condition resource and will return Empty list if condition is not found for the given name
```
Did you confirm that it actually returns an empty list instead of null?

corneliouzbett replied:
I am not doing any search for this ticket
https://issues.openmrs.org/projects/FM/issues/FM-65

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl2_2.java:
```diff
@@ -0,0 +1,143 @@
+	@Override
+	public Condition createCondition(Condition condition) {
+		if (condition.isEmpty()) {
+			throw new UnprocessableEntityException();
```
Can you add the error message in the constructor()?

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl2_2.java:
```diff
@@ -0,0 +1,143 @@
+	}
+
+	/**
+	 * @see org.openmrs.module.fhir.api.helper.ConditionHelper#createCondition(Condition)
```
Can you include the package of the `Condition` in the above `see` annotation parameter?

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl2_2.java:
```diff
@@ -0,0 +1,143 @@
+public class ConditionHelperImpl2_2 implements ConditionHelper {
+
+	/**
+	 * @see org.openmrs.module.fhir.api.helper.ConditionHelper#getCondition(String)
```
Can you include the package of the above `String`?

dkayiwa commented on api-2.2/src/main/java/org/openmrs/module/fhir/helper/ConditionHelperImpl2_2.java:
```diff
@@ -0,0 +1,143 @@
+	}
+
+	/**
+	 * @see org.openmrs.module.fhir.api.helper.ConditionHelper#generateOpenMrsCondition(Condition)
```
Can you include the package of the `Condition` in the above `see` annotation parameter?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionHelperImpl.java:
```diff
@@ -0,0 +1,75 @@
+public class BaseConditionHelperImpl {
+
+	/**
+	 * @see org.openmrs.module.fhir.api.helper.ConditionHelper#getCondition(String)
```
Can you include the package of the `String` in the above `see` annotation parameter?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/impl/base/BaseConditionHelperImpl.java:
```diff
@@ -0,0 +1,75 @@
+	}
+
+	/**
+	 * @see org.openmrs.module.fhir.api.helper.ConditionHelper#createCondition(Condition)
```
Can you include the package of the `Condition` in the above `see` annotation parameter?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/ConditionStrategy.java:
```diff
@@ -0,0 +1,58 @@
+public class ConditionStrategy implements GenericConditionStrategy {
+
+	/**
+	 * @see org.openmrs.module.fhir.api.strategies.condition.GenericConditionStrategy#getConditionById(String)
```
Can you include the package of the `String` in the above `see` annotation parameter?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/ConditionStrategy.java:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	/**
+	 * @see org.openmrs.module.fhir.api.strategies.condition.GenericConditionStrategy#searchConditionById(String)
```
Can you include the package of the `String` in the above `see` annotation parameter?

dkayiwa commented on api/src/main/java/org/openmrs/module/fhir/api/strategies/condition/ConditionStrategy.java:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	/**
+	 * @see org.openmrs.module.fhir.api.strategies.condition.GenericConditionStrategy#createFHIRCondition(Condition)
```
Can you include the package of the `Condition` in the above `see` annotation parameter?

dkayiwa commented on omod/src/main/java/org/openmrs/module/fhir/providers/RestfulConditionResourceProvider.java:
```diff
@@ -13,17 +13,16 @@
-import ca.uhn.fhir.rest.annotation.Read;
-import ca.uhn.fhir.rest.annotation.RequiredParam;
-import ca.uhn.fhir.rest.annotation.Search;
+import ca.uhn.fhir.rest.annotation.*;
```
https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotusewildcardimports

jonathandick commented:
Well done @corneliouzbett ! Thank you @dkayiwa for all the attention and feedback paid to this PR!
